### PR TITLE
Debounce autocomplete tags to boost performance by reducing redundant calls to database

### PR DESCRIPTION
**Pull request by aayushman-singh**
Solves #12415. Adding simple server side results limiter to 10 and a debouncer for client side for now.

### CODE REVIEW

File: wagtail/admin/static_src/wagtailadmin/js/vendor/tag-it.js [hunk 1]
```diff
@@ -316,6 +328,15 @@
         _tags: function() {
             return this.tagList.find('.tagit-choice:not(.removed)');
         },
+        
+        // Debouncer for tags that accepts a delay arg
+        debounce: function(func, delay) {
```
Comment (laymonage) [inline wagtail/admin/static_src/wagtailadmin/js/vendor/tag-it.js:333]: We already have a debounce util in https://github.com/wagtail/wagtail/blob/main/client/src/utils/debounce.ts

File: wagtail/admin/static_src/wagtailadmin/js/vendor/tag-it.js [hunk 2]
```diff
@@ -280,29 +280,41 @@
                         that.createTag(that._cleanedInput());
                     }
                 });
-
-            // Autocomplete.
+    
+            //Autocomplete
```
Comment (laymonage) [inline wagtail/admin/static_src/wagtailadmin/js/vendor/tag-it.js:284]: Please avoid unnecessary changes

**Review by laymonage — COMMENTED** [review]
Thank you for giving this a go! This will need tests, particularly the Python changes

File: wagtail/admin/static_src/wagtailadmin/js/vendor/tag-it.js (hunk 1 above)
Comment (aayushman-singh) [reply wagtail/admin/static_src/wagtailadmin/js/vendor/tag-it.js:333]: Oh, I don't know how I missed that, thanks!

**Comment by aayushman-singh** [comment]
Noted, I'll be sure to test it thoroughly before opening for review.

**Comment by aayushman-singh** [comment]
https://github.com/user-attachments/assets/15e151ac-dc6f-4ac1-bcdd-938c951fd4a6

A delay of 2000 was used to clearly show the debouncer working, actual code uses 400. 25 tags were added for trying it out, as seen in the video it shows top 10 only.

File: wagtail/admin/views/tags.py [hunk 3]
```diff
@@ -19,7 +19,7 @@ def autocomplete(request, app_name=None, model_name=None):
 
     term = request.GET.get("term", None)
     if term:
-        tags = tag_model.objects.filter(name__istartswith=term).order_by("name")
+        tags = tag_model.objects.filter(name__istartswith=term).order_by("name")[:10]
```
Comment (lb-) [inline wagtail/admin/views/tags.py:22]: Can you please also add Python tests for this.

I would suggest we include a comment as to why we have filtered these here.

```suggestion
        # limit the tags provided to the client for performance when projects have a large amount of tags
        tags = tag_model.objects.filter(name__istartswith=term).order_by("name")[:10]
```

File: client/src/controllers/TagController.ts [hunk 4]
```diff
@@ -20,17 +21,31 @@ export class TagController extends Controller {
...
+      },
+      this.debounceValue,
+    );
 
     $(this.element).tagit({
-      autocomplete: { source: this.urlValue },
+      autocomplete: { source: debouncedAutocomplete },
       preprocessTag,
       ...this.optionsValue,
     });
```
Comment (lb-) [inline client/src/controllers/TagController.ts:51]:
```suggestion
    });
}

autocomplete(request: { term: string }, response: (data: any) => void) {
        $.ajax({...});

```

File: client/src/controllers/TagController.ts [hunk 5]
```diff
@@ -20,17 +21,31 @@ export class TagController extends Controller {
   static values = {
     options: { default: {}, type: Object },
     url: String,
+    debounce: { type: Number, default: 400 },
```
Comment (lb-) [inline client/src/controllers/TagController.ts:24]: Let's default this to 300 as you have in your comment below

```suggestion
    debounce: { type: Number, default: 300 },
```

File: client/src/controllers/TagController.test.js [hunk 6]
```diff
@@ -58,8 +58,6 @@ describe('TagController', () => {
     const [{ preprocessTag }] = tagitMock.mock.calls[0];
 
     expect(preprocessTag).toBeInstanceOf(Function);
-
-    expect(preprocessTag()).toEqual();
```
Comment (lb-) [inline client/src/controllers/TagController.test.js:62]: We will need more tests added please, mocking the API request/respsonse via mocking the ajax result.

File: client/src/controllers/TagController.ts [hunk 7]
```diff
@@ -20,17 +21,31 @@ export class TagController extends Controller {
...
   };
 
   declare optionsValue: any;
   declare urlValue: any;
+  declare debounceValue: number;
   tagit?: JQuery<HTMLElement>;
 
   connect() {
     const preprocessTag = this.cleanTag.bind(this);
+    // Debounces by specified debounceValue to reduce unnecessary calls to database while user types, default is 300
```
Comment (lb-) [inline client/src/controllers/TagController.ts:34]: I would recommend we remove this comment, it's better to just look where the default is set instead of re-writing what the default is.

As for below, let's follow the patterns in the other controllers where we abstract the method out and create a debounced wrapper around it in the connect method.


```
  this.autocomplete = debounce(this.autocomplete).bind(this);
```

File: client/src/controllers/TagController.ts [hunk 8]
```diff
@@ -20,17 +21,31 @@ export class TagController extends Controller {
...
+    const debouncedAutocomplete = debounce(
+      (request: { term: string }, response: (data: any) => void) => {
+        $.ajax({
```
Comment (lb-) [inline client/src/controllers/TagController.ts:37]: While this file does use a jQuery component, I think we should be careful to be adding more jQuery usage, could you please look at using `fetch` instead.

I know the syntax is a bit more verbose but it means that we can more easily migrate this to a non-jQuery solution in the future.

**Review by lb- — CHANGES_REQUESTED** [review]
Thank you @aayushman-singh  this is an awesome start.

There's a bit of work needed before this can be ready for a full review & potential merge, comments have been added.

You may want to prepare two separate PRs, one for the Python/Django changes and another for the JS/Stimulus changes.

This way you can focus on tests for things in isolation, up to you though.

Unit tests are super critical for any change, especially for things like this where we want to ensure that we continue to keep these improvements long term. Often the code change itself can be super small but writing the tests can take a lot more time.

Finally, this code does not seem to consider 'in flight' requests, this would be built in for the existing code but if we control the fetching we would need to consider this and ensure we have a way to cancel existing requests that come back 'after' others.

File: client/src/controllers/TagController.test.js (hunk 6 above)
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:62]: I'll work on that, not really confident in testing with python but I'll do my best

File: client/src/controllers/TagController.ts (hunk 5 above)
Comment (aayushman-singh) [reply client/src/controllers/TagController.ts:24]: Alright, whatever feels better

File: client/src/controllers/TagController.ts (hunk 7 above)
Comment (aayushman-singh) [reply client/src/controllers/TagController.ts:34]: Noted

File: client/src/controllers/TagController.ts (hunk 8 above)
Comment (aayushman-singh) [reply client/src/controllers/TagController.ts:37]: I initially started out with fetch but due to the overwhelming usage of jQuery everywhere in the code base I didn't want to change things unnecessarily. I'll switch to using fetch.

File: wagtail/admin/views/tags.py (hunk 3 above)
Comment (aayushman-singh) [reply wagtail/admin/views/tags.py:22]: Yes, I'll look into python testing for this. Thanks

**Comment by aayushman-singh** [comment]
Since javascript and Stimulus tag has been applied to this PR I'll be moving the python changes to another branch as advised. I'll look into the in-flight requests as well. Thanks for your guidance.

**Comment by aayushman-singh** [comment]
@lb- could we close this soon? I feel like it might be ready to merge.

File: client/src/controllers/TagController.ts [hunk 9]
```diff
@@ -20,22 +21,64 @@ export class TagController extends Controller {
...
+  declare urlValue: string;
+  declare debounceValue: number;
+  private currentAbortController: AbortController | null = null;
   tagit?: JQuery<HTMLElement>;
 
   connect() {
     const preprocessTag = this.cleanTag.bind(this);
 
+    // Ensure `this` is bound to the correct context in `autocomplete`
+    const debouncedAutocomplete = debounce(
```
Comment (lb-) [inline client/src/controllers/TagController.ts:37]: I'd suggest we just call this variable `autocomplete`. 

```suggestion
    const autocomplete = debounce(
```

Typescript will tell us what it is if needed.

**Review by lb- — COMMENTED** [review]
This is looking really solid from a code perspective, I've yet to.run this locally but here's some feedback so far. 

I'll try to do a full review locally soon.

File: client/src/controllers/TagController.ts [hunk 10]
```diff
@@ -20,22 +21,64 @@ export class TagController extends Controller {
...
+        method: 'GET',
+        headers: {
+          Accept: 'application/json',
+        },
+        signal,
+      });
+
+      const data = await fetchResponse.json();
+      response(Array.isArray(data) ? data : []);
+    } catch (error) {
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.ts:75]: Will do.

File: client/src/controllers/TagController.test.js [hunk 11]
```diff
@@ -25,52 +25,156 @@ describe('TagController', () => {
...
+    await controller.autocomplete({ term: 'error' }, response);
+
+    expect(global.fetch).toHaveBeenCalledWith(
+      '/admin/tag-autocomplete/?term=error',
+      expect.any(Object),
+    );
+    expect(response).toHaveBeenCalledWith([]);
+  });
+
+  it('should call the function only once after debounce delay', () => {
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:135]: Ok, I'll remove this one

File: client/src/controllers/TagController.test.js [hunk 12]
```diff
@@ -25,52 +25,156 @@ describe('TagController', () => {
...
+      <form id="form">
+        <input
+          id="id_tags"
+          type="text"
+          name="tags"
+          data-controller="w-tag"
+          data-action="example:event->w-tag#clear"
+          data-w-tag-options-value="{&quot;allowSpaces&quot;:true,&quot;tagLimit&quot;:10}"
+          data-w-tag-url-value="/admin/tag-autocomplete/"
+        >
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:51]: Understood.

File: client/src/controllers/TagController.ts [hunk 13]
```diff
@@ -20,22 +21,64 @@ export class TagController extends Controller {
...
-  declare urlValue: any;
+  declare urlValue: string;
+  declare debounceValue: number;
+  private currentAbortController: AbortController | null = null;
   tagit?: JQuery<HTMLElement>;
 
   connect() {
     const preprocessTag = this.cleanTag.bind(this);
 
+    // Ensure `this` is bound to the correct context in `autocomplete`
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.ts:36]: Ok.

File: client/src/controllers/TagController.ts (hunk 9 above)
Comment (aayushman-singh) [reply client/src/controllers/TagController.ts:37]: Ok.

File: client/src/controllers/TagController.test.js [hunk 14]
```diff
@@ -25,52 +25,156 @@ describe('TagController', () => {
...
+      }),
+    );
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
   });
 
   it('should attach the jQuery tagit to the controlled element', async () => {
+    jest.useRealTimers(); // Switch to real timers for this test
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:40]: I tried using `await jest.resolveAllTimersAsync()`, but it doesn't handle `requestAnimationFrame` properly since it relies on real browser timing. Jest's mocked timers like `jest.runAllTimers()` or `jest.resolveAllTimersAsync()` only simulate time-based operations like `setTimeout` and `setInterval`, which don't cover `requestAnimationFrame`. This test depends on real timing for browser-like interactions, so `jest.useRealTimers()` is necessary. However, feel free to double-check if there's an alternative I might have missed.

File: client/src/controllers/TagController.test.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:51]: I'm sorry, does 'this indentation' mean my changes or the previous code?

File: client/src/controllers/TagController.test.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/TagController.test.js:39]: Hmm ok. I'll take another look. Thanks for explaining.

File: client/src/controllers/TagController.test.js [hunk 15]
```diff
@@ -25,52 +25,156 @@ describe('TagController', () => {
...
       preprocessTag: expect.any(Function),
       tagLimit: 10,
     });
 
     expect(element[0]).toEqual(document.getElementById('id_tags'));
 
-    // check the supplied preprocessTag function
     const [{ preprocessTag }] = tagitMock.mock.calls[0];
-
     expect(preprocessTag).toBeInstanceOf(Function);
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:67]: I understand but the source is no longer a static string since I changed it to an object because of the new autocomplete function using fetch. Is it alright if I change it to use an object?

[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/TagController.test.js:68]: I am not sure I follow, why would the autocomplete functionality (requesting tags to use as you type) impact the pre-processing of tags?

I might misunderstand the logic, but I would expect the pre-processing for free-tagging to work the same.

Remember, the user can enter any value, not just what's available on the server.

https://docs.wagtail.org/en/stable/reference/pages/model_recipes.html#disabling-free-tagging

[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:68]: Thank you for raising this concern! My changes only updated the source property in `autocomplete` to fetch suggestions dynamically via fetch. This affects how suggestions are retrieved, but it does not impact the preprocessTag functionality or the free-tagging behavior.

Users can still enter any value they like, whether or not it matches the server-provided suggestions, and the preprocessTag logic (e.g., quoting tags with spaces) continues to work as before.

The purpose of the change was solely to make autocomplete suggestions dynamic, while ensuring free-tagging and preprocessing remain consistent with their original behavior. But its possible I haven't understood your point completely, please let me know if my explanation doesn't makes sense.

**Review by lb- — COMMENTED** [review]
I have more suggestions coming soon, just getting some testing running locally, here's the fixes for the existing unit tests.

File: client/src/controllers/TagController.test.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:50]: Ok.

File: client/src/controllers/TagController.test.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:68]: It seems my comment did not reach when I planned it to, please ignore since you have gone to the trouble of writing it yourself

[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js]: Thank you for helping out here, I'm sorry for causing so much trouble. I understand of course that comments should be left as is and that the main goal is to do the minimum viable changes to existing tests to support the new behavior. I have made the suggested changes and committed them. Please let me know if anything else pops up.

File: client/src/controllers/TagController.test.js [hunk 16]
```diff
@@ -1,5 +1,4 @@
 import $ from 'jquery';
-
 import { Application } from '@hotwired/stimulus';
 import { TagController } from './TagController';
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:6]: Sounds good

File: client/src/controllers/TagController.test.js [hunk 17]
```diff
@@ -1,5 +1,4 @@
 import $ from 'jquery';
-
 import { Application } from '@hotwired/stimulus';
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:4]: Done in latest commit.

**Comment by aayushman-singh** [comment]
Thank you for your detailed review and assistance with the tests.

File: client/src/controllers/TagController.test.js [hunk 18]
```diff
@@ -69,8 +83,98 @@ describe('TagController', () => {
...
+      1,
+      '/admin/tag-autocomplete/?term=test1',
+      expect.any(Object),
+    );
+    expect(fetch).toHaveBeenNthCalledWith(
+      2,
+      '/admin/tag-autocomplete/?term=test2',
+      expect.any(Object),
+    );
+  });
```
Comment (lb-) [inline client/src/controllers/TagController.test.js:179]: @aayushman-singh  here's a revised approach to the tests, instead of pulling in the controller and calling the methods directly, we should do our best to test the application facing behaviour.

This is a bit tricky for the tagit widget but here's a suggested refined approach.

The core difference is instead of doing this.

```
    const controller = new TagController();
    controller.urlValue = '/admin/tag-autocomplete/'; // Set urlValue manually
    const response = jest.fn();
    await controller.autocomplete({ term: 'nonexistent' }, response);
```

We do this

```js
  describe('autocomplete', () => {
    let autocompleteSource;

    beforeAll(async () => {
      document.body.innerHTML = `
      <form id="form">
        <input
          id="id_tags"
          type="text"
          name="tags"
          data-controller="w-tag"
          data-action="example:event->w-tag#clear"
          data-w-tag-options-value='${JSON.stringify({
            allowSpaces: true,
            tagLimit: 10,
          })}'
          data-w-tag-url-value="/admin/tag-autocomplete/"
        >
      </form>`;

      await flushPromises();

      autocompleteSource = tagitMock.mock.calls[0][0].autocomplete.source;
    });

   // then in each test we have access to `autocompleteSource` as it would have been called on tagit
      const result = await new Promise((resolve) => {
        autocompleteSource({ term: 'tag' }, resolve);
        jest.runAllTimers();
      });
```

See what you think I know this is a sizeable re-write but hopefully that's ok and still has the main tests running.

If you are happy with this, please review locally & commit, feel free to refine of course.

```js
import $ from 'jquery';

import { setImmediate } from 'timers';
import { Application } from '@hotwired/stimulus';
import { TagController } from './TagController';

window.$ = $;

jest.useFakeTimers();

const flushPromises = () => new Promise(setImmediate);

describe('TagController', () => {
  let application;
  let element;

  const tagitMock = jest.fn(function innerFunction() {
    element = this;
  });

  window.$.fn.tagit = tagitMock;

  element = null;

  beforeAll(() => {
    application = Application.start();
    application.register('w-tag', TagController);
    application.handleError = jest.fn();
  });

  beforeEach(() => {
    element = null;
    jest.clearAllMocks();
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve([]),
      }),
    );
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should attach the jQuery tagit to the controlled element', async () => {
    document.body.innerHTML = `
  <form id="form">
    <input
      id="id_tags"
      type="text"
      name="tags"
      data-controller="w-tag"
      data-action="example:event->w-tag#clear"
      data-w-tag-options-value="{&quot;allowSpaces&quot;:true,&quot;tagLimit&quot;:10}"
      data-w-tag-url-value="/admin/tag-autocomplete/"
    >
  </form>`;

    expect(tagitMock).not.toHaveBeenCalled();

    await flushPromises();

    expect(tagitMock).toHaveBeenCalledWith({
      allowSpaces: true,
      autocomplete: { source: expect.any(Function) },
      preprocessTag: expect.any(Function),
      tagLimit: 10,
    });

    expect(element[0]).toEqual(document.getElementById('id_tags'));

    // check the supplied preprocessTag function
    const [{ preprocessTag }] = tagitMock.mock.calls[0];

    expect(preprocessTag).toBeInstanceOf(Function);

    expect(preprocessTag()).toEqual();
    expect(preprocessTag('"flat white"')).toEqual(`"flat white"`);
    expect(preprocessTag("'long black'")).toEqual(`"'long black'"`);
    expect(preprocessTag('caffe latte')).toEqual(`"caffe latte"`);

    // check the custom clear behavior
    document
      .getElementById('id_tags')
      .dispatchEvent(new CustomEvent('example:event'));

    await flushPromises();

    expect(tagitMock).toHaveBeenCalledWith('removeAll');
  });

  describe('autocomplete', () => {
    let autocompleteSource;

    beforeAll(async () => {
      document.body.innerHTML = `
      <form id="form">
        <input
          id="id_tags"
          type="text"
          name="tags"
          data-controller="w-tag"
          data-action="example:event->w-tag#clear"
          data-w-tag-options-value='${JSON.stringify({
            allowSpaces: true,
            tagLimit: 10,
          })}'
          data-w-tag-url-value="/admin/tag-autocomplete/"
        >
      </form>`;

      await flushPromises();

      autocompleteSource = tagitMock.mock.calls[0][0].autocomplete.source;
    });

    it('should handle a successful fetch response', async () => {
      global.fetch = jest.fn(() =>
        Promise.resolve({
          json: () => Promise.resolve(['tag1', 'tag2', 'tag3']),
        }),
      );

      expect(autocompleteSource).toBeInstanceOf(Function);

      const result = await new Promise((resolve) => {
        autocompleteSource({ term: 'tag' }, resolve);
        jest.runAllTimers();
      });

      expect(result).toEqual(['tag1', 'tag2', 'tag3']);

      expect(global.fetch).toHaveBeenCalledTimes(1);
      expect(global.fetch).toHaveBeenCalledWith(
        expect.stringContaining('/admin/tag-autocomplete/?term=tag'),
        {
          headers: { Accept: 'application/json' },
          method: 'GET',
          signal: expect.any(AbortSignal),
        },
      );
    });

    it('should handle an empty fetch response', async () => {
      global.fetch = jest.fn(() =>
        Promise.resolve({
          json: () => Promise.resolve([]),
        }),
      );

      expect(autocompleteSource).toBeInstanceOf(Function);

      const result = await new Promise((resolve) => {
        autocompleteSource({ term: 'nonexistent' }, resolve);
        jest.runAllTimers();
      });

      expect(global.fetch).toHaveBeenCalledWith(
        'http://localhost/admin/tag-autocomplete/?term=nonexistent',
        expect.any(Object),
      );

      expect(result).toEqual([]);
    });

    it('should handle fetch errors gracefully', async () => {
      global.fetch = jest.fn(() => Promise.reject(new Error('Fetch error')));

      expect(application.handleError).not.toHaveBeenCalled();
      expect(autocompleteSource).toBeInstanceOf(Function);

      const result = await new Promise((resolve) => {
        autocompleteSource({ term: 'error' }, resolve);
        jest.runAllTimers();
      });

      expect(global.fetch).toHaveBeenCalledWith(
        'http://localhost/admin/tag-autocomplete/?term=error',
        expect.any(Object),
      );

      // check we returned an empty array, even with an error
      expect(result).toEqual([]);

      // check the error gets handed to Stimulus
      expect(application.handleError).toHaveBeenCalledWith(
        new Error('Fetch error'),
        'Failed to fetch autocomplete tags.',
        { term: 'error', url: '/admin/tag-autocomplete/' },
      );
    });

    it('should debounce requests to the tag endpoint', async () => {
      // mocking a slow fetch
      global.fetch = jest.fn(
        () =>
          new Promise((resolve) => {
            window.setTimeout(resolve, 300);
          }),
      );

      expect(autocompleteSource).toBeInstanceOf(Function);

      const response = jest.fn();

      // First request
      autocompleteSource({ term: 'test1' }, response);
      jest.advanceTimersByTime(100);

      // Second request during debounce window
      autocompleteSource({ term: 'test2' }, response);
      jest.advanceTimersByTime(100);

      // Third request during debounce window
      autocompleteSource({ term: 'test3' }, response);
      jest.advanceTimersByTime(50);

      await jest.runAllTimersAsync();

      expect(fetch).toHaveBeenCalledTimes(1);

      expect(fetch).toHaveBeenCalledWith(
        // only the most recent request should be made
        expect.stringContaining('/admin/tag-autocomplete/?term=test3'),
        expect.any(Object),
      );

      expect(response).toHaveBeenCalledTimes(1);
      expect(response).toHaveBeenCalledWith([]);
    });

    it('should allow new requests to to be made after the debounce window & abort in-flight', async () => {
      // mock a slow fetch the first time but fast the second time
      global.fetch = jest
        .fn()
        .mockImplementationOnce(
          () =>
            new Promise((resolve) => {
              window.setTimeout(
                () =>
                  resolve({ json: () => Promise.resolve(['slow IGNORED']) }),
                600,
              );
            }),
        )
        .mockImplementationOnce(
          () =>
            new Promise((resolve) => {
              window.setTimeout(
                () => resolve({ json: () => Promise.resolve(['slow cooked']) }),
                50,
              );
            }),
        );

      expect(autocompleteSource).toBeInstanceOf(Function);

      const response = jest.fn();

      // First request
      autocompleteSource({ term: 's' }, response);
      jest.advanceTimersByTime(300);

      // Second request after previous debounce window
      autocompleteSource({ term: 'slow' }, response);

      await jest.runAllTimersAsync();

      expect(fetch).toHaveBeenCalledTimes(2);

      expect(fetch).toHaveBeenNthCalledWith(
        1,
        expect.stringContaining('/admin/tag-autocomplete/?term=s'),
        expect.any(Object),
      );

      expect(fetch).toHaveBeenNthCalledWith(
        2,
        expect.stringContaining('/admin/tag-autocomplete/?term=slow'),
        expect.any(Object),
      );

      expect(application.handleError).not.toHaveBeenCalled();

      expect(response).toHaveBeenCalledTimes(2);
    });
  });
});
```

File: client/src/controllers/TagController.ts [hunk 19]
```diff
@@ -1,6 +1,7 @@
 import $ from 'jquery';
```
Comment (lb-) [inline client/src/controllers/TagController.ts:2]: For the `TagController`, this is great work and functional, but here's a bit of a refined approach.

Instead of setting up the debounced function in `connect`, let's create a new method on `initialize` and use it directly, calling it `autocompleteLazy` (we have a similar convention in other files, see `SwapController`.

In addition, instead of passing the `response` callback to the `autocomplete` method, this is already a promise so we can make it simply return a value, and then that's the resolve callback.

We end up with something like this.

```ts
  initialize() {
    this.autocompleteLazy = debounce(
      this.autocomplete.bind(this),
      this.debounceValue,
    );
  }

  connect() {
    const preprocessTag = this.cleanTag.bind(this);

    const autocomplete = {
      source: (request, response) => {
        this.autocompleteLazy.cancel();
        this.autocompleteLazy(request).then(response);
      },
    };

    $(this.element).tagit({
      autocomplete,
      preprocessTag,
      ...this.optionsValue,
    });
  }
```

The other changes below are mostly just variable names, refining the error handling and using the built in browser functionality to generate the URL (instead of writing strings manually, it's better to use `URL` & `searchParams`.

A reminder that technically the urlValue can be blank, but we do not want to default this to the known API key, instead we want this to error out. So putting the `new URL` inside the `try/catch` block means that empty values will be automatically handled for us.

Have a look at the full code below, see what you think, if it works locally and you are happy with this approach, please push these changes in your next commits.


```ts
import $ from 'jquery';

import { Controller } from '@hotwired/stimulus';
import { debounce } from '../utils/debounce';

declare global {
  interface JQuery {
    tagit: (options: Record<string, any> | string) => void;
  }
}

/**
 * Attach the jQuery tagit UI to the controlled element.
 *
 * See https://github.com/aehlke/tag-it
 *
 * @example
 * <input id="id_tags" type="text" name="tags" data-controller="w-tag" data-w-tag-url-value="/admin/tag-autocomplete/" />
 */
export class TagController extends Controller {
  static values = {
    debounce: { type: Number, default: 300 },
    options: { default: {}, type: Object },
    url: { default: '', type: String },
  };

  declare optionsValue: any;
  declare urlValue: string;
  declare debounceValue: number;

  private autocompleteAbort: AbortController | null = null;
  private autocompleteLazy;

  tagit?: JQuery<HTMLElement>;

  initialize() {
    this.autocompleteLazy = debounce(
      this.autocomplete.bind(this),
      this.debounceValue,
    );
  }

  connect() {
    const preprocessTag = this.cleanTag.bind(this);

    const autocomplete = {
      source: (request, response) => {
        this.autocompleteLazy.cancel();
        this.autocompleteLazy(request).then(response);
      },
    };

    $(this.element).tagit({
      autocomplete,
      preprocessTag,
      ...this.optionsValue,
    });
  }

  async autocomplete({ term }: { term: string }) {
    if (this.autocompleteAbort) {
      this.autocompleteAbort.abort();
    }

    this.autocompleteAbort = new AbortController();
    const { signal } = this.autocompleteAbort;

    try {
      const url = new URL(this.urlValue, window.location.origin);
      url.searchParams.set('term', term);

      const fetchResponse = await fetch(url.toString(), {
        headers: { Accept: 'application/json' },
        method: 'GET',
        signal,
      });

      const data = await fetchResponse.json();
      return Array.isArray(data) ? data : [];
    } catch (error) {
      // Allow API errors to pass through, ensure they are logged.
      this.context.application.handleError(
        error,
        'Failed to fetch autocomplete tags.',
        { term, url: this.urlValue },
      );
    } finally {
      this.autocompleteAbort = null;
    }

    return [];
  }

  /**
   * Double quote a tag if it contains a space
   * and if it isn't already quoted.
   */
  cleanTag(val: string) {
    return val && val[0] !== '"' && val.indexOf(' ') > -1 ? `"${val}"` : val;
  }

  /**
   * Method to clear all the tags that are set.
   */
  clear() {
    $(this.element).tagit('removeAll');
  }
}
```

**Review by lb- — CHANGES_REQUESTED** [review]
OK, I have some larger suggestions, I hope this is OK.

Trying to help polish this off and ensure that we have tests/code set up for future enhancements.

See how you go with this suggested approach. Thank you for the incredible work here so far.

File: client/src/controllers/TagController.test.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/TagController.test.js]: You're not causing any trouble at all, this is a bit complex and I'm also learning myself as I go. 

I am thankful you've gone to the huge effort of working through the PR, finding a direction to head and taking on feedback.

File: client/src/controllers/TagController.ts [hunk 20]
```diff
@@ -20,22 +21,73 @@ export class TagController extends Controller {
   static values = {
     options: { default: {}, type: Object },
     url: String,
+    debounce: { type: Number, default: 300 },
```
[the remark this answers is not in the document]
Comment (aayushman-singh) [reply client/src/controllers/TagController.ts:24]: Will do, not a problem at all.

File: client/src/controllers/TagController.test.js (hunk 18 above)
Comment (aayushman-singh) [reply client/src/controllers/TagController.test.js:179]: Alright, this is not passing all tests currently, I'll make sure everything works locally and make some changes and let you know.

File: client/src/controllers/TagController.ts (hunk 19 above)
Comment (aayushman-singh) [reply client/src/controllers/TagController.ts:2]: Thanks, will do.

**Comment by aayushman-singh** [comment]
Added a test to check graceful handling of empty urls as you have set up `try/catch` for the same, though it might be overkill. I have also slightly modified the `catch` condition to differentiate errors more precisely and used the same for tests. 

Everything has been tested locally and all tests are passing with the new approach. Let me know if my modifications were unneccessary, I think this should be ready to merge @lb- . Thanks for the assist!

**Comment by lb-** [comment]
Epic, I will try to have a look in the next few days, thank you!

File: client/src/controllers/TagController.test.js [hunk 21]
```diff
@@ -64,13 +79,218 @@ describe('TagController', () => {
...
+            allowSpaces: true,
+            tagLimit: 10,
+          })}'
+          data-w-tag-url-value="/admin/tag-autocomplete/"
+        >
+      </form>`;
+
+      await flushPromises();
+
+      autocompleteSource = tagitMock.mock.calls[0][0].autocomplete.source;
```
Comment (lb-) [inline client/src/controllers/TagController.test.js:114]: This had a small issue in it, we should be getting the last call, as it's run in previous tests.

```suggestion
      // get last mock call
      autocompleteSource =
        tagitMock.mock.calls[tagitMock.mock.calls.length - 1][0].autocomplete
          .source;
          
```

File: client/src/controllers/TagController.ts [hunk 22]
```diff
@@ -18,24 +19,88 @@ declare global {
  */
 export class TagController extends Controller {
   static values = {
+    delay: { type: Number, default: 300 },
```
Comment (lb-) [inline client/src/controllers/TagController.ts:22]: Let's default this to zero (`0`) and then apply the value of `200` at the widget level, using `300` is a bit too slow.

I can add this when merging though + adjust the unit test.

File: client/src/controllers/TagController.ts [hunk 23]
```diff
@@ -18,24 +19,88 @@ declare global {
...
+        headers: { Accept: 'application/json' },
+        method: 'GET',
+        signal,
+      });
+
+      const data = await fetchResponse.json();
+      return Array.isArray(data) ? data : [];
+    } catch (error) {
+      const errorMessage =
+        error instanceof DOMException && error.name === 'AbortError'
```
Comment (lb-) [inline client/src/controllers/TagController.ts:89]: No need to log the AbortError, this is not really an error as it's intentional, let's just return when this happens.

**Review by lb- — APPROVED** [review]
Thank you @aayushman-singh I have tested this in a few browsers and with a few scenarios, including slow network connections.

This is a vast improvement and should avoid us thrashing the server when user types.

I will adjust a few minor things, the main one being setting the default `delay` to zero and then setting this to `200` at the `AdminTagWidget` instead. This aligns with a lot of our other defaults that are 'falsey' and then let the values be set as needed.

Amazing work here, thanks for taking on the feedback and doing the heavy lifting on this PR.
